utils/typename-derive/src/lib.rs

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro::{self, TokenStream};
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput};
+
+#[proc_macro_derive(TypeName)]
+pub fn derive(input: TokenStream) -> TokenStream {
+    let DeriveInput { ident, .. } = parse_macro_input!(input);
+    let type_name = ident.to_string();
+    let output = quote! {
+        impl TypeName for #ident {
+            fn typename_str() -> &'static str {
+                #type_name
+            }
+        }
+    };
+    output.into()
```

Comment:
Should generics be handled somehow? I.e. should `<Vec<Foo>>::typename_str()` be `Vec<Foo>` or just `Vec`?

Reply:
I actually tested it and it seems to not work at all. I would prefer it to be `Vec<Foo>`. I'll have to research how to do this.